crates/red_knot_python_semantic/docs/kind_of_types.md

```diff
@@ -0,0 +1,201 @@
+
+A sealed type in Python has the property
+that its proper subtypes[^2] are all disjunct from one another,
+and the union of all the proper subtypes of a sealed type is exactly equal to the sealed type.
+
+All singleton types are sealed types; however, not all sealed types are singleton types.
+
+Examples of sealed types (other than the singleton types listed above) are:
+
+- `bool`:
+
+    - The only inhabitants of `bool` at runtime are the constants `True` and `False`.
+    - The only proper subtypes of `bool` are `Literal[True]`, `Literal[False]`, and `Never`.
+        `Literal[True]`, `Literal[False]` and `Never` are all disjunct types,
+        and their union is exactly equal to `bool`.
+
+- Enums: consider the following enum class:
+
+    ```py
+    from enum import Enum, auto
+
+    class Foo(Enum):
+        X = auto()
+        Y = auto()
+        Z = auto()
+    ```
+
+    - The only inhabitants of the `Foo` type at runtime are the enum `Foo`'s members:
+        `Foo.X` and `Foo.Y`
+    - The only proper subtypes of `Foo` are `Literal[Foo.X]`, `Literal[Foo.Y]`, `Literal[Foo.Z]`, and `Never`.
+        `Literal[Foo.X]`, `Literal[Foo.Y]`, `Literal[Foo.Z]` and `Never` are all disjunct,
+        and the union `Literal[Foo.X, Foo.Y, Foo.Z] | Never` is exactly equal to the type `Foo`.
+
+Because a sealed type is equivalent to the union of all of its proper subtypes,
+for any given sealed type `X` where the only proper subtypes of `X` are `A`, `B` and `C`,
+`X & ~A == B | C`. To give a Python example:
+
+```py
+from enum import Enum, auto
+
+class X(Enum):
+    A = auto()
+    B = auto()
+    C = auto()
+
+
+def f(var: X):
+    ...  # the type of `var` here is `X` (which is equivalent to `Literal[X.A, X.B, X.C]`)
+
+    if var is X.A:
+        ...  # var is narrowed to `Literal[X.A]`
+    else:
+        ...  # var is narrowed to `Literal[X.B, X.C]`
+```
+
+## Single-value types
+
+A single-value type is a non-empty type for which it is known
+that all inhabitants of the type are equivalent with respect to their runtime value.
+All singleton types are single-value types, but not all single-value types are singleton types.
+
+Examples of single-value types that are not singleton types
+are `Literal["foo"]`, `Literal[b"foo"]`, `Literal[123456]`, and `tuple[Literal[True], Literal[False]]`.
+All runtime inhabitants of the `Literal["foo"]` type are entirely fungible and equal
+(they are all equal to the string `"foo"`).
+However, they are not necessarily the *same* object in terms of identity;
+multiple instances of the `"foo"` string can exist at runtime at different memory addresses.[^1]
+This means that it is not safe to narrow a single-value type by identity unless it is also known
+that the type is a singleton type.
+
+In order for a type to be considered a single-value type,
+there must exist a reflexive, symmetric, and transitive equivalence relation
+between all inhabitants of the type. It follows that for a single-value type in Python,
```

Comment:
The definition of a single-value type is specific to the Python equality relation, so it seems odd to say "there must exist a ... relation" here.

I would say instead that a single-value type must have a reflexive, symmetric, and transitive definition of Python equality.

But really I kind of think this entire paragraph might not be necessary; I think the literal-type examples you give above are probably sufficient to clarify what a real single-value type looks like. In practice we are never going to try to define any single-value type on a non-builtin class, or any class that might challenge the requirements.

Reply:
> `Literal[E.A]`, for an enum `E` with a member `A`, is a single-value type.

Fair! And if we observe a custom `__eq__` implementation on an `Enum` subclass, we probably have to stop treating literal types of those enum members as single-valued types, because we no longer can claim to understand the equality relation for instances of that enum type.